packages/lodestar/src/tasks/tasks/archiveBlocks.ts

```diff
@@ -4,81 +4,60 @@
 
 import {ITask} from "../interface";
 import {IBeaconDb} from "../../db/api";
-import {SignedBeaconBlock} from "@chainsafe/lodestar-types";
+import {Checkpoint} from "@chainsafe/lodestar-types";
 import {IBeaconConfig} from "@chainsafe/lodestar-config";
 import {ILogger} from "@chainsafe/lodestar-utils/lib/logger";
-import {toHexString} from "@chainsafe/ssz";
-import {BlockSummary} from "../../chain";
-import {computeEpochAtSlot} from "@chainsafe/lodestar-beacon-state-transition";
+import {ForkChoice} from "@chainsafe/lodestar-fork-choice";
+import {computeStartSlotAtEpoch} from "@chainsafe/lodestar-beacon-state-transition";
 
 export interface IArchiveBlockModules {
   db: IBeaconDb;
+  forkChoice: ForkChoice;
   logger: ILogger;
 }
 
 /**
  * Archives finalized blocks from active bucket to archive bucket.
  */
 export class ArchiveBlocksTask implements ITask {
+  private readonly config: IBeaconConfig;
   private readonly db: IBeaconDb;
+  private readonly forkChoice: ForkChoice;
   private readonly logger: ILogger;
-  private readonly config: IBeaconConfig;
 
-  private finalized: BlockSummary;
-  private pruned: BlockSummary[];
+  private finalized: Checkpoint;
 
-  public constructor(
-    config: IBeaconConfig,
-    modules: IArchiveBlockModules,
-    finalized: BlockSummary,
-    pruned: BlockSummary[]
-  ) {
+  public constructor(config: IBeaconConfig, modules: IArchiveBlockModules, finalized: Checkpoint) {
+    this.config = config;
     this.db = modules.db;
+    this.forkChoice = modules.forkChoice;
     this.logger = modules.logger;
-    this.config = config;
     this.finalized = finalized;
-    this.pruned = pruned;
   }
 
   /**
    * Only archive blocks on the same chain to the finalized checkpoint.
    */
   public async run(): Promise<void> {
     this.logger.profile("Archive Blocks");
-    const allBlockEntries = await this.db.block.entries();
-    const blockEntries = allBlockEntries.filter(({value}) => value.message.slot <= this.finalized.slot);
-    const blocksByRoot = new Map<string, SignedBeaconBlock>(
-      blockEntries.map(({key, value}) => [toHexString(key), value])
-    );
-
-    const archivedBlocks: SignedBeaconBlock[] = [];
-    const finalizedBlock = blocksByRoot.get(toHexString(this.finalized.blockRoot));
-    if (finalizedBlock) {
-      archivedBlocks.push(finalizedBlock);
-    }
-    let lastBlock = finalizedBlock;
-    while (lastBlock) {
-      lastBlock = blocksByRoot.get(toHexString(lastBlock.message.parentRoot));
-      if (lastBlock) {
-        archivedBlocks.push(lastBlock);
+    const finalizedSlot = computeStartSlotAtEpoch(this.config, this.finalized.epoch);
+    const keysToDelete = [];
+    let totalArchived = 0;
+    for await (const {key, value} of await this.db.block.entriesStream()) {
+      if (value.message.slot > finalizedSlot) {
+        continue;
       }
+      if (this.forkChoice.isDescendant(key, this.finalized.root)) {
+        await this.db.blockArchive.add(value);
```

Comment:
sometimes it comes up to 3.5s to finish the task: `2020-09-12 12:31:37  [CHORES]           info: Archive Blocks - duration=3509ms`. Maybe just collect all and do `batchAdd` to save time? If it's too much then probably collect and save per n blocks.

Reply:
I refactored this task bc it caused issues archiving the medalla chasm. Pulling all blocks into memory could cause OOM. The slowdown here is the tradeoff of not pulling all blocks in memory at once.
Unfortunately, collecting all and doing a batchAdd will have the same issue as pulling all into memory at once.

Collecting N blocks at a time seems like a good middle ground. I'll make an issue for this as this PR is getting too big.